### PR TITLE
add examples/A2SamlWithAzureAD.md

### PR DESCRIPTION
**Pull request by kenmacleod**
HowTo - Setting up SAML Auth in A2 with Azure AD

### CODE REVIEW

File: examples/A2SamlWithAzureAD.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+   1. Azure AD Identifier.
+   1. Logout URL.
+
+#### Chef Automate V2
+
+1. SSH to your A2 instance.
+2. Create a file called `saml.toml` where we will put in configuration information.
+3. In the `saml.toml` file, you will need to add the following information in the example format listed:
+   1. `ca_contents`: this is the X.509 Certificate from step 5 above. Copy this and paste as the value for ca_contents. Use three double quotes to indicate a multiline string. (ex: `"""`)
+   1. `sso_url`: This is the value for Identity Provider Single Sign-On URL using the Login URL Value from step 6 above.
```
Comment (msorens) [inline examples/A2SamlWithAzureAD.md]:
```suggestion
   1. `sso_url`: This is the value for Identity Provider Single Sign-On URL using the Login URL value from step 6 above.
```

File: examples/A2SamlWithAzureAD.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+   1. `email_attr`: The value for this should be `emailaddress`
+   1. `username_attr`: The value for this should be `username`
+   1. `entity_issuer`: This should be the URL for your automate server plus `/dex/callback`. (ex: https://serverfqdn/dex/callback)
+   1. Save and close the file.
```
Comment (msorens) [inline examples/A2SamlWithAzureAD.md]: This is a real cliffhanger. 😁 You should explain how to apply `saml.toml` to the system, unless that is done elsewhere.

File: examples/A2SamlWithAzureAD.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+6. Note the following values from Section 4:
+   1. Login URL.
+   1. Azure AD Identifier.
+   1. Logout URL.
+
+#### Chef Automate V2
+
+1. SSH to your A2 instance.
+2. Create a file called `saml.toml` where we will put in configuration information.
+3. In the `saml.toml` file, you will need to add the following information in the example format listed:
```
Comment (msorens) [inline examples/A2SamlWithAzureAD.md]:
```suggestion
3. In the `saml.toml` file, add the following information in the example format listed:
```

File: examples/A2SamlWithAzureAD.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#### References
+- https://www.brittanynwoods.com/blog/automate-2-saml
+
+#### Azure AD Config
+1. In azure ad portal (https://portal.azure.com) under azure active directory and then under enterprise apps. Create a new Non-Gallery Application
+2. After naming it, Click on single signon and select SAML
+3. In section 1 – Basic SAML Configuration, enter following:
+   1. Identifier (Entity ID): https://serverfqdn/dex/callback
+   1. Reply URL (Assertion Consumer Service URL): https://serverfqdn/dex/callback
+4. In section 2: User Attributes and Claims check the defaults that are there.
```
Comment (msorens) [inline examples/A2SamlWithAzureAD.md]:
```suggestion
4. In section 2: User Attributes and Claims, check the defaults that are there.
```

File: examples/A2SamlWithAzureAD.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+##### Versions Tested On
+- Chef Automate | [2.x]
+
+#### References
+- https://www.brittanynwoods.com/blog/automate-2-saml
+
+#### Azure AD Config
+1. In azure ad portal (https://portal.azure.com) under azure active directory and then under enterprise apps. Create a new Non-Gallery Application
+2. After naming it, Click on single signon and select SAML
```
Comment (msorens) [inline examples/A2SamlWithAzureAD.md]:
```suggestion
2. After naming it, click on single sign-on and select SAML
```

**Comment by jmassardo** [comment]
Thanks for the example. I was able to use it to get auth working in my lab. You might also consider adding `name_id_policy_format = "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress"` to the bottom of your toml example. This will allow AAD to return the proper ID that IAMv2 needs to set permissions and views.

**Comment by danielcbright** [comment]
@kenmacleod I would love to merge this, can you address the issues listed in the comments? once those are done, I'll go ahead and pull this in, thanks for the contribution!!!

**Comment by danielcbright** [comment]
I'm going to add the last bit to it, it's just one command to apply the toml

**Comment by srenatus** [comment]
I'm having concerns about signing key rotation. In step 5, we're retrieving the signing cert, and later set it in A2. That is the correct approach in general, however,

- Azure AD [does signing key rotation](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-signing-key-rollover), and the way this is conveyed to clients (i.e. A2 in this case) is via [SAML IdP Metadata](https://docs.microsoft.com/en-us/azure/active-directory/azuread-dev/azure-ad-federation-metadata#token-signing-certificates).

    > A federation metadata document published by Azure AD can have multiple signing keys, such as **when Azure AD is preparing to update the signing certificate**. When a federation metadata document includes more than one certificate, a service that is validating the tokens should support all certificates in the document.
- When copying the signing cert over, it won't ever be updated.
- A2 **cannot** consume SAML metadata, and this **cannot** keep track of those signing certs on its own.
- It seems like Azure has provided [some scripts to help with manual cert rotation](https://github.com/AzureAD/azure-activedirectory-powershell-tokenkey), but I cannot tell if they are of use for us.

It's been a while since I've last poked at one of those signing certs, but I think we'd do good to check their expiry date! I'm unaware of that realistic expectations should be around key rotation schedules, but if one happens, it'll prohibit all SAML users from signin in.

**Pull request merged**